espnet/nets/pytorch_backend/e2e_tts_fastspeech.py

```diff
@@ -0,0 +1,623 @@
+        # bias init
+        for p in self.parameters():
+            if p.dim() == 1:
+                p.data.zero_()
+        # reset some modules with default init
+        for m in self.modules():
+            if isinstance(m, (torch.nn.Embedding, LayerNorm)):
+                m.reset_parameters()
+
+    @property
+    def attention_plot_class(self):
+        return TTSPlot
+
+    @property
+    def base_plot_keys(self):
+        """base key names to plot during training. keys should match what `chainer.reporter` reports
+
+        if you add the key `loss`, the reporter will report `main/loss` and `validation/main/loss` values.
+        also `loss.png` will be created as a figure visulizing `main/loss` and `validation/main/loss` values.
+
+        :rtype list[str] plot_keys: base keys to plot during training
+        """
+        plot_keys = ["loss", "l1_loss", "duration_loss"]
+        if self.use_scaled_pos_enc:
+            plot_keys += ["encoder_alpha", "decoder_alpha"]
+
+        return plot_keys
+
+
+class DurationCalculator(torch.nn.Module):
+    """Duration calculator using teacher model
+
+    :param e2e_tts_transformer.Transformer teacher_model: teacher auto-regressive Transformer
+    """
+
+    def __init__(self, teacher_model):
+        super(DurationCalculator, self).__init__()
+        if not isinstance(teacher_model, Transformer):
+            raise ValueError("teacher model should be the instance of e2e_tts_transformer.Transformer")
+        self.teacher_model = teacher_model
+
+    def forward(self, xs, ilens, ys, olens):
+        """Calculate duration of each inputs
+
+        :param torch.Tensor xs: batch of padded character ids (B, Tmax)
+        :param torch.Tensor ilens: list of lengths of each input batch (B)
+        :param torch.Tensor ys: batch of padded target features (B, Lmax, odim)
+        :param torch.Tensor ilens: list of lengths of each output batch (B)
+        :return torch.Tensor: batch of durations (B, Tmax)
+        """
+        att_ws = self._calculate_attentions(xs, ilens, ys, olens)
+        # TODO(kan-bayashi): fix this issue
+        # this does not work in multi-gpu case. registered buffer is not saved.
+        if not hasattr(self, "diag_head_idx"):
+            self._init_diagonal_head(att_ws)
+        att_ws = att_ws[:, self.diag_head_idx]
+        durations = [self._calculate_duration(att_w, ilen, olen) for att_w, ilen, olen in zip(att_ws, ilens, olens)]
+
+        return pad_list(durations, 0)
+
+    @staticmethod
+    def _calculate_duration(att_w, ilen, olen):
+        return torch.stack([att_w[:olen, :ilen].argmax(-1).eq(i).sum() for i in range(ilen)])
+
+    def _init_diagonal_head(self, att_ws):
+        diagonal_scores = att_ws.max(dim=-1)[0].mean(dim=-1).mean(dim=0)  # (H * L,)
+        self.register_buffer("diag_head_idx", diagonal_scores.argmax())
+
+    def _calculate_attentions(self, xs, ilens, ys, olens):
```

Comment:
I think this function depends too much on `self.teacher_model` (e.g., no access to `self.xxx` but `self.teacher_model`, and access private methods in `teacher_model`). How about moving this method to `teacher_model` instead?

Reply:
I agree with you.
I'm curious that unnecessary calculations in `teacher_model.calculate_all_attentions()` can be bottleneck.
If not so, I will replace this function with `calculate_all_attentions`.

Reply:
You can create `Transformer.calculate_all_attentions(..., skip_xxx=True)` or  `Transformer.calculate_teacher_attentions(...)`. I'm just afraid to export `_xxx` methods to this external class.

Reply:
fixed by adding new options in `calculate_all_attentions`